meeshkan/serve/mock/storage/entity.py

```diff
@@ -0,0 +1,206 @@
+        res = {}
+        for method_name in self.methods:
+            method: Operation = getattr(path_item, method_name)
+            if (
+                method is not None
+                and method._x is not None
+                and "x-meeshkan-operation" in method._x
+            ):
+                if (
+                    method._x["x-meeshkan-operation"] == "insert"
+                    or method._x["x-meeshkan-operation"] == "upsert"
+                ):
+                    request_body = typing.cast(RequestBody, method.requestBody)
+                    if "application/json" in request_body.content:
+                        schema = request_body.content["application/json"].schema
+                        if schema is not None:
+                            entity_path = self._find_entity(schema, "$")
+                            if res is not None:
+                                res[method_name] = parse(entity_path)
+        return res
+
+    def _find_entity(self, schema, path):
+        if isinstance(schema, Reference):
+            name = schema._ref.split("/")[3]
+            if name == self._entity_name:
+                return path
+        elif schema.get("type", None) == "array":
+            if "items" not in schema:
+                return None
+            elif isinstance(schema["items"], list):
+                return next(
+                    (
+                        x
+                        for x in (
+                            self._find_entity(item, path + "[*]")
+                            for item in schema["items"]
+                        )
+                        if x is not None
+                    )
+                )
+            else:
+                return self._find_entity(schema["items"], path + "[*]")
+        elif schema.get("type", "object") == "object" and "properties" in schema:
+            return next(
+                (
+                    x
+                    for x in (
+                        self._find_entity(
+                            schema["properties"][p], "{}.{}".format(path, p)
+                        )
+                        for p in schema["properties"]
+                    )
+                    if x is not None
+                )
+            )
+
+        return None
+
+    def filter(self, request: Request):
+        return lambda x: True
+
+    def id_filter(self, request: Request):
+        return None
+
+    def extract_id(self, request: Request):
+        match = _match_to_path(request.pathname, self._pathname)
+        if match is None or "id" not in match:
+            return self.id_filter(request)
+
+        return match["id"]
+
+    def extract_entity(self, request: Request):
+        found = self._request_entity_selectors[request.method.value].find(
+            request.bodyAsJson
+        )
+        if len(found) == 0:
+            return None
+        return found[0].value
+
+
+class Entity:
```

Comment:
I don't understand/remember anymore what entities are. What are example entities? What are they used for? Do they keep some storage?

Reply:
I've added docstrings. Maybe, they can help. An entity is a data d structure extracted from an API interface and linked to a specific places in requests and responses. So it can be posted and retrieved automatically. And it can be accessed from a callback.